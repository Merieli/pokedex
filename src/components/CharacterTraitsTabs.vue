<script setup lang="ts">
import { ref } from 'vue';

import { Episode, LocationCharacter } from '@/domain';

interface Props {
    gender: string;
    origin: string;
    type: string;
    location: LocationCharacter;
    lastEpisode: Episode;
    episode: Episode[];
}

defineProps<Props>();

const tab = ref(null);
</script>

<template>
    <v-card class="character-traits-tabs__sheet">
        <v-tabs v-model="tab" class="character-traits-tabs__tabs" color="primary" align-tabs="center">
            <v-tab value="about">About</v-tab>
            <v-tab value="episodes">Episodes</v-tab>
        </v-tabs>

        <v-card-text class="character-traits-tabs__sheet-content">
            <v-window v-model="tab">
                <v-window-item
                    class="character-traits-tabs__tab-content"
                    value="about"
                    data-character-traits-tabs="about"
                >
                    <v-list lines="five">
                        <div class="character-traits-tabs__content-first">
                            <v-list-item class="character-traits-tabs__item" title="Gender" :subtitle="gender">
                                <template #prepend>
                                    <v-icon color="blue">mdi-gender-male-female</v-icon>
                                </template>
                            </v-list-item>

                            <v-list-item class="character-traits-tabs__item" title="Born in" :subtitle="origin">
                                <template #prepend>
                                    <v-icon color="red">mdi-heart-pulse</v-icon>
                                </template>
                            </v-list-item>
                        </div>
                        <v-list-item class="character-traits-tabs__item" title="Type" :subtitle="type">
                            <template #prepend>
                                <v-icon color="purple">mdi-account-search</v-icon>
                            </template>
                        </v-list-item>
                        <v-list-item
                            class="character-traits-tabs__item"
                            title="Location"
                            data-character-traits-tabs="location"
                        >
                            <template #prepend>
                                <v-icon color="green">mdi-earth-box</v-icon>
                            </template>

                            <template #subtitle>
                                <p>Name: {{ location.name }}</p>
                                <p>Dimension: {{ location.dimension }}</p>
                                <p>Type: {{ location.type }}</p>
                            </template>
                        </v-list-item>
                        <v-list-item
                            class="character-traits-tabs__item"
                            title="Last episode"
                            data-character-traits-tabs="last-episode"
                        >
                            <template #prepend>
                                <v-icon color="black">mdi-video-box</v-icon>
                            </template>
                            <template #subtitle>
                                <p>Name: {{ lastEpisode.name }}</p>
                                <p>Season: {{ lastEpisode.season }}</p>
                                <p>Episode: {{ lastEpisode.number }}</p>
                            </template>
                        </v-list-item>
                    </v-list>
                </v-window-item>

                <v-window-item class="character-traits-tabs__tab-content" value="episodes">
                    <v-chip
                        v-for="(currentEpisode, index) in episode"
                        :key="index"
                        class="character-traits-tabs__episode ma-2"
                    >
                        S{{ currentEpisode.season }} - E{{ currentEpisode.season }}: {{ currentEpisode.name }} [{{
                            currentEpisode.air_date
                        }}]
                    </v-chip>
                </v-window-item>
            </v-window>
        </v-card-text>
    </v-card>
</template>

<style lang="postcss" scoped>
.character-traits-tabs {
    &__sheet {
        @apply rounded-t-3xl
            w-full max-w-5xl
            h-[40%]
            fixed bottom-0;

        @media (min-width: theme('screens.laptop')) {
            margin-left: calc((100% - theme('screens.laptop')) / 2);
            margin-right: calc((100% - theme('screens.laptop')) / 2);
        }
    }

    &__sheet-content {
        @apply overflow-y-scroll;

        &.v-card-text {
            @apply min-h-fit max-h-[80%];
        }

        .v-window__container {
            @apply mb-12;
        }
    }

    &__tabs {
        @apply text-meri-low
            pt-8 h-auto;
    }

    &__content-first {
        display: flex;
        @apply flex justify-between flex-col mobile:flex-row;

        .character-traits-tabs__item {
            @apply w-full mobile:w-1/2;
        }
    }

    &__item {
        @apply mb-5;

        p {
            @apply mb-2;
        }

        .v-list-item__prepend {
            @apply items-start h-full mr-2 mobile:mr-4;
        }
    }

    &__tab-content {
        @apply px-0 mobile:px-10 
            mb-3;
    }
}
</style>
<style scoped>
.character-traits-tabs .v-list-item__prepend > .v-icon {
    margin-right: 8px;
}
</style>
