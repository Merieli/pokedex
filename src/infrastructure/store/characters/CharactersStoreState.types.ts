export interface CharactersStoreState {
    characters: any[];
}
