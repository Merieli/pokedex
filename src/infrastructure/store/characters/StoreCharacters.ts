import { defineStore } from 'pinia';
import { CharactersStoreState } from './CharactersStoreState.types';
import getAllCharacters from './actions/getAllCharacters';

const state = (): CharactersStoreState => ({
    characters: [],
});

const getters = {};

const actions = {
    getAllCharacters,
};

export const useCounterStore = defineStore('characters', {
    state,
    actions,
});
