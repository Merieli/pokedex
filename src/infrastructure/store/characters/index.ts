export * from './CharactersStoreState.types';
export * from './StoreCharacters';
