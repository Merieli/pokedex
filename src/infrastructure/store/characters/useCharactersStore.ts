import { computed, reactive, Ref, ref } from 'vue';

import {
    ApiCharacter,
    Character,
    StorePaginationCharacter,
    StoreRandomCharacter,
    StoreSearchCharacter,
} from '@domain/index';
import { defineStore } from 'pinia';

import { CharacterDTO } from '@/domain/character/CharacterDTO';
import { logError } from '@/helpers/logError';
import CharactersGatewayHttp from '@/infrastructure/gateway/CharactersGatewayHttp';

export const useCharactersStore = defineStore('characters', () => {
    const api = new CharactersGatewayHttp();

    const allCharacters: Ref<Character[]> = ref([]);
    const charactersPerPage: Ref<Record<number, Character[]>> = ref({});
    const selectedCharacter: Character = reactive({
        id: '',
        name: '',
        episode: [],
        gender: 'unknown',
        status: 'unknown',
        type: '',
        image: '',
        species: 'unknown',
        location: {
            id: '',
            name: '',
            dimension: '',
            type: '',
        },
        origin: '',
    });

    const random: StoreRandomCharacter = reactive({
        character: [],
        show: false,
    });
    const search: StoreSearchCharacter = reactive({
        text: '',
        characters: [],
        pagination: {
            currentPage: 1,
            total: 0,
            results: 0,
        },
    });
    const pagination: StorePaginationCharacter = reactive({
        currentPage: 1,
        total: 0,
        results: 0,
    });
    const isLoading: Ref<boolean> = ref(false);
    const isSelected: Ref<boolean> = ref(false);

    const isSearching = computed(() => search.characters && search.characters.length > 0);

    const prepareCharacterData = (listOfCharacters: ApiCharacter[]): Character[] => {
        const newList: Character[] = listOfCharacters.map((character) => new CharacterDTO(character));
        return newList;
    };

    /**
     * Action to get all characters of api and save in state
     */
    const getAllCharacters = async (): Promise<void> => {
        try {
            isLoading.value = true;
            const response = await api.getAll(pagination.currentPage);

            const listOfCharactersPrepared = prepareCharacterData(response.results);

            Object.defineProperty(charactersPerPage.value, pagination.currentPage, {
                value: listOfCharactersPrepared,
                writable: false,
                enumerable: true,
                configurable: true,
            });

            const newList = allCharacters.value.concat(listOfCharactersPrepared);
            allCharacters.value = newList;

            pagination.total = response.info.pages;
            pagination.results = response.info.count;
        } catch (error) {
            const { code } = logError('Action getAllCharacters', error);
            // notifyUser(code, 'imprimir todos os personagens')  //função para exibir o erro para o usuário
        } finally {
            isLoading.value = false;
        }
    };

    /**
     * Search for a character by name in the api and store it in the store
     * @param name - character's name
     */
    const findCharacterByName = async (name: string) => {
        try {
            isLoading.value = true;
            search.text = name;
            const response = await api.findByName(name, 1);
            const listOfCharactersPrepared = prepareCharacterData(response.results);

            search.characters = listOfCharactersPrepared;
            search.pagination.total = response.info.pages;
            search.pagination.results = response.info.count;
        } catch (error) {
            const { code } = logError('Action findCharacterByName', error);
        } finally {
            isLoading.value = false;
        }
    };

    /**
     * Searches for a random character and saves it in the store by activating its display
     */
    const generateRandomCharacter = async () => {
        try {
            isLoading.value = true;
            const integerRandomNumberUpTo826 = Math.floor(Math.random() * 826) + 1;

            const response = await api.findByIds(integerRandomNumberUpTo826);
            const listOfCharactersPrepared = prepareCharacterData(response);

            random.character = listOfCharactersPrepared;
            random.show = true;
        } catch (error) {
            const { code } = logError('Action generateRandomCharacter', error);
        } finally {
            isLoading.value = false;
        }
    };

    return {
        allCharacters,
        charactersPerPage,
        selectedCharacter,
        search,
        pagination,
        random,
        isLoading,
        isSearching,
        isSelected,
        getAllCharacters,
        findCharacterByName,
        generateRandomCharacter,
    };
});
