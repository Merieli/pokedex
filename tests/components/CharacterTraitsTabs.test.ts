import CharacterTraitsTabs from '@components/CharacterTraitsTabs.vue';
import { createTestingPinia } from '@pinia/testing';
import { mount } from '@vue/test-utils';
import { Pinia } from 'pinia';
import { createVuetify } from 'vuetify';

import { mockSelectedCharacter } from '../__mocks__/mockSelectedCharacter';

import { useCharactersStore } from '@/infrastructure/store/characters';

describe('CharacterTraitsTabs.test.vue', () => {
    let pinia: Pinia;

    const setupWrapper = () => {
        pinia = createTestingPinia({
            initialState: {
                characters: {
                    selectedCharacter: mockSelectedCharacter,
                    isSelected: true,
                },
            },
        });

        const vuetify = createVuetify();

        return {
            wrapper: mount(CharacterTraitsTabs, {
                global: {
                    plugins: [pinia, vuetify],
                    stubs: {
                        teleport: true,
                    },
                },
            }),
            store: useCharactersStore(),
        };
    };

    describe('IntegraÃ§Ã£o/Componente', () => {
        describe('ðŸ‘€ RenderizaÃ§Ã£o:', () => {
            test('Dado o componente Quando renderizado EntÃ£o deve possuir html equivalente ao Snapshot gravado', () => {
                const { wrapper } = setupWrapper();

                expect(wrapper.html()).toMatchSnapshot();
            });

            test('Dado uma lista de caraterÃ­sticas do personagem Quando renderizado EntÃ£o deve exibir nas abas os dados da localizaÃ§Ã£o do personagem', () => {
                const { wrapper } = setupWrapper();
                const location = wrapper.find('[data-character-traits-tabs="location"]');

                expect(location.exists()).toBeTruthy();
            });

            test('Dado uma lista de caraterÃ­sticas do personagem Quando renderizado EntÃ£o deve exibir nas abas os dados do Ãºltimo episÃ³dio com o personagem', () => {
                const { wrapper } = setupWrapper();
                const location = wrapper.find('[data-character-traits-tabs="last-episode"]');

                expect(location.exists()).toBeTruthy();
            });
        });
    });
});
