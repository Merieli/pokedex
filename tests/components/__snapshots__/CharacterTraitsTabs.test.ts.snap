// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CharacterTraitsTabs.test.vue > Integra√ß√£o/Componente > üëÄ Renderiza√ß√£o: > Dado o componente Quando renderizado Ent√£o deve possuir html equivalente ao Snapshot gravado 1`] = `
"<v-card class=\\"character-traits-tabs__sheet\\" data-v-ca6c6aea=\\"\\">
  <v-tabs class=\\"character-traits-tabs__tabs\\" color=\\"primary\\" align-tabs=\\"center\\" data-v-ca6c6aea=\\"\\">
    <v-tab value=\\"about\\" data-v-ca6c6aea=\\"\\">About</v-tab>
    <v-tab value=\\"episodes\\" data-v-ca6c6aea=\\"\\">Episodes</v-tab>
  </v-tabs>
  <v-card-text class=\\"character-traits-tabs__sheet-content\\" data-v-ca6c6aea=\\"\\">
    <v-window data-v-ca6c6aea=\\"\\">
      <v-window-item class=\\"character-traits-tabs__tab-content\\" data-character-traits-tabs=\\"about\\" value=\\"about\\" data-v-ca6c6aea=\\"\\">
        <v-list lines=\\"five\\" data-v-ca6c6aea=\\"\\">
          <v-row no-gutters=\\"\\" data-v-ca6c6aea=\\"\\">
            <v-col cols=\\"5\\" data-v-ca6c6aea=\\"\\">
              <v-list-item class=\\"character-traits-tabs__item\\" title=\\"Gender\\" data-v-ca6c6aea=\\"\\"></v-list-item>
            </v-col>
            <v-col cols=\\"7\\" data-v-ca6c6aea=\\"\\">
              <v-list-item class=\\"character-traits-tabs__item\\" title=\\"Born in\\" data-v-ca6c6aea=\\"\\"></v-list-item>
            </v-col>
          </v-row>
          <v-list-item class=\\"character-traits-tabs__item\\" title=\\"Type\\" data-v-ca6c6aea=\\"\\"></v-list-item>
          <v-list-item class=\\"character-traits-tabs__item\\" title=\\"Location\\" data-character-traits-tabs=\\"location\\" data-v-ca6c6aea=\\"\\"></v-list-item>
          <v-list-item class=\\"character-traits-tabs__item\\" title=\\"Last episode\\" data-character-traits-tabs=\\"last-episode\\" data-v-ca6c6aea=\\"\\"></v-list-item>
        </v-list>
      </v-window-item>
      <v-window-item class=\\"character-traits-tabs__tab-content\\" value=\\"episodes\\" data-v-ca6c6aea=\\"\\"></v-window-item>
    </v-window>
  </v-card-text>
</v-card>"
`;
